### PR TITLE
Added EuroSat dataset

### LINKED ISSUE #166: Add the EuroSat dataset
EuroSat seems like a good dataset to add for a few reasons:
- It is included in the Tensorflow datasets (https://www.tensorflow.org/datasets/catalog/eurosat)
- It is one of the "specialized" remote sensing datasets included in the VTAB set (https://arxiv.org/pdf/1910.04867v2.pdf). VTAB has been used in several recent popular papers e.g. MLP-Mixer, Big Transfer.
- It would make @isaaccorley's spectral index notebook more simple

### PR DESCRIPTION
**Pull request by calebrob6**
closes #166

### CODE REVIEW

**Comment by calebrob6** [comment]
Anyone know why mypy wants `sample["label"]` to be a `Tensor` here in eurosat.py but is cool with `sample["label"]` being an `int` in so2sat.py? I think having the label as a Tensor will break training scripts.

https://github.com/microsoft/torchgeo/blob/main/torchgeo/datasets/so2sat.py#L131

File: tests/datasets/test_eurosat.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (adamjstewart) [inline tests/datasets/test_eurosat.py:1]: Were `AnnualCrop_1.tif` and `Forest_1.tif` intentionally added or should they be in the zip file?

File: torchgeo/datasets/eurosat.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+import rasterio
+import torch
+from torch import Tensor
+
+from .geo import VisionDataset
+from .utils import check_integrity, download_and_extract_archive
+
+
+class EuroSat(VisionDataset):
+    r"""EuroSat dataset.
```
Comment (adamjstewart) [inline torchgeo/datasets/eurosat.py:20]: Does this need to be a raw string? I don't see any backslashes in here

File: torchgeo/datasets/eurosat.py [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+    a total of 27,000 labeled and geo-referenced images.
+
+    Dataset format:
+
+    * rasters are 13-channel GeoTiffs
+    * labels are values in the range [0,9]
+
+    Dataset classes:
+
+    1. Industrial Buildings
```
Comment (adamjstewart) [inline torchgeo/datasets/eurosat.py:33]: Should this be numbered 0–9?

File: torchgeo/datasets/eurosat.py [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    * https://ieeexplore.ieee.org/document/8736785
+    * https://ieeexplore.ieee.org/document/8519248
+    """
+
+    url = "http://madm.dfki.de/files/sentinel/EuroSATallBands.zip"  # 2.0 GB download
+    filename = "EuroSATallBands.zip"
+    md5 = "5ac12b3b2557aa56e1826e981e8e200e"
+
+    # for some reason the class directories are acutally nested in this directory
+    base_dir = "ds/images/remote_sensing/otherDatasets/sentinel_2/tif"
```
Comment (adamjstewart) [inline torchgeo/datasets/eurosat.py:55]: This should use `os.path.join` since the path separator differs depending on OS

File: tests/datasets/test_eurosat.py (hunk 1 above)
Comment (calebrob6) [reply tests/datasets/test_eurosat.py:1]: They are fake test data and are in the zip file as well

File: torchgeo/datasets/eurosat.py (hunk 3 above)
Comment (calebrob6) [reply torchgeo/datasets/eurosat.py:33]: It doesn't match up with the class numbers in the dataset or anything, just a listing of classes

**Comment by isaaccorley** [comment]
Just wanted to mention that the PatternNet and RESISC45 datasets return the label as tensors so if that breaks anything downstream, as you pointed out, we will need to change them as well.

File: tests/datasets/test_eurosat.py (hunk 1 above)
Comment (adamjstewart) [reply tests/datasets/test_eurosat.py:1]: But do they need to be outside of the zip file too?

Comment (calebrob6) [reply tests/datasets/test_eurosat.py:1]: I don't think so, but in other datasets we have added the fake data outside of the zip

Comment (isaaccorley) [reply tests/datasets/test_eurosat.py:1]: In the more recent classification datasets I didn't keep the data outside the zip.

**Comment by calebrob6** [comment]
I removed the fake data outside the zip file because it is unnecessary

File: tests/datasets/test_eurosat.py (hunk 1 above)
Comment (adamjstewart) [reply tests/datasets/test_eurosat.py:1]: The reason that some datasets have fake data outside the zip is that those datasets have trainers we need to test. Of course we could refactor things to copy and extract the zip files, but that's why things are the way they are.

**Comment by calebrob6** [comment]
We should force the trainers to download the dataset if it doesn't exist

File: torchgeo/datasets/eurosat.py [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+    * https://ieeexplore.ieee.org/document/8736785
+    * https://ieeexplore.ieee.org/document/8519248
+    """
+
+    url = "http://madm.dfki.de/files/sentinel/EuroSATallBands.zip"  # 2.0 GB download
+    filename = "EuroSATallBands.zip"
+    md5 = "5ac12b3b2557aa56e1826e981e8e200e"
+
+    # for some reason the class directories are acutally nested in this directory
```
Comment (adamjstewart) [inline torchgeo/datasets/eurosat.py:54]:
```suggestion
    # For some reason the class directories are actually nested in this directory
```

File: torchgeo/datasets/eurosat.py [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+
+    def __len__(self) -> int:
+        """Return the number of data points in the dataset.
+
+        Returns:
+            length of the dataset
+        """
+        return len(self.labels)
+
+    @lru_cache()
```
Comment (adamjstewart) [inline torchgeo/datasets/eurosat.py:154]: Does this need to be cached? I don't think the cache will be hit very often since it only contains the most recent 128 images and the total dataset size is 27K. So it may actually slow things down due to overhead.

Comment (calebrob6) [reply torchgeo/datasets/eurosat.py:154]: I don't think it matters

**Review by adamjstewart — APPROVED** [review]
